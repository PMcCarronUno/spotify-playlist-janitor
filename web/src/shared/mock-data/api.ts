import { LoginResponse } from "api/auth-api/auth-api-types";
import { DatabasePlaylistResponse, SkippedTrackResponse } from "api/data-api";
import { SpotifyPlaylistResponse } from "api/spotify-api";

type AuthData = {
  login: LoginResponse;
};

export const auth: AuthData = {
  login: {
    refresh_token: "access_token",
    token_type: "Bearer",
    access_token: "access_token",
    expires_in: 4503,
  },
};

export const databasePlaylists: DatabasePlaylistResponse[] = [
  {
    id: "1",
    skipThreshold: 10,
    ignoreInitialSkips: false,
    autoCleanupLimit: 10,
  },
  {
    id: "2",
    skipThreshold: 2,
    ignoreInitialSkips: false,
    autoCleanupLimit: 2,
  },
  {
    id: "3",
    skipThreshold: 3,
    ignoreInitialSkips: true,
    autoCleanupLimit: 3,
  },
  {
    id: "4",
    skipThreshold: 4,
    ignoreInitialSkips: true,
    autoCleanupLimit: 4,
  },
];

export const spotifyPlaylists: SpotifyPlaylistResponse[] = [
  {
    id: "1",
    name: "Playlist 1",
    href: "https://open.spotify.com/playlist/1",
    images: [
      {
        height: 640,
        width: 640,
        url: "https://mosaic.scdn.co/101",
      },
      {
        height: 300,
        width: 300,
        url: "https://mosaic.scdn.co/102",
      },
      {
        height: 60,
        width: 60,
        url: "https://mosaic.scdn.co/103",
      },
    ],
  },
  {
    id: "2",
    name: "Playlist 2",
    href: "https://open.spotify.com/playlist/2",
    images: [
      {
        height: 640,
        width: 640,
        url: "https://mosaic.scdn.co/201",
      },
      {
        height: 300,
        width: 300,
        url: "https://mosaic.scdn.co/202",
      },
      {
        height: 60,
        width: 60,
        url: "https://mosaic.scdn.co/203",
      },
    ],
  },
  {
    id: "3",
    name: "Playlist 3",
    href: "https://open.spotify.com/playlist/3",
    images: [
      {
        height: 640,
        width: 640,
        url: "https://mosaic.scdn.co/301",
      },
      {
        height: 300,
        width: 300,
        url: "https://mosaic.scdn.co/302",
      },
      {
        height: 60,
        width: 60,
        url: "https://mosaic.scdn.co/303",
      },
    ],
  },
  {
    id: "4",
    name: "Playlist 4",
    href: "https://open.spotify.com/playlist/4",
    images: [
      {
        height: 640,
        width: 640,
        url: "https://mosaic.scdn.co/401",
      },
      {
        height: 300,
        width: 300,
        url: "https://mosaic.scdn.co/402",
      },
      {
        height: 60,
        width: 60,
        url: "https://mosaic.scdn.co/403",
      },
    ],
  },
  {
    id: "5",
    name: "Playlist 5",
    href: "https://open.spotify.com/playlist/5",
    images: [
      {
        height: 640,
        width: 640,
        url: "https://mosaic.scdn.co/501",
      },
      {
        height: 300,
        width: 300,
        url: "https://mosaic.scdn.co/502",
      },
      {
        height: 60,
        width: 60,
        url: "https://mosaic.scdn.co/503",
      },
    ],
  },
  {
    id: "6",
    name: "Playlist 6",
    href: "https://open.spotify.com/playlist/6",
    images: [
      {
        height: 640,
        width: 640,
        url: "https://mosaic.scdn.co/601",
      },
      {
        height: 300,
        width: 300,
        url: "https://mosaic.scdn.co/602",
      },
      {
        height: 60,
        width: 60,
        url: "https://mosaic.scdn.co/603",
      },
    ],
  },
];

export const skippedTracks: SkippedTrackResponse = [
  {
    playlistId: databasePlaylists[0].id,
    trackId: "track 01",
    name: "Name 1",
    skippedDate: new Date("2023-10-31T19:51:25Z"),
    artists: [
      {
        id: "artist 01",
        name: "Artist 1",
        href: "https://open.spotify.com/artist/1",
      }
    ],
    album: {
      id: "album 01",
      name: "Album 1",
      href: "https://open.spotify.com/album/1",
      images: [
        {
          height: 240,
          width: 240,
          url: "https://mosaic.scdn.co/101",
        }
      ]
    }
  },
  {
    playlistId: databasePlaylists[0].id,
    trackId: "track 01",
    name: "Name 1",
    skippedDate: new Date("2023-10-31T19:51:38Z"),
    artists: [
      {
        id: "artist 01",
        name: "Artist 1",
        href: "https://open.spotify.com/artist/1",
      }
    ],
    album: {
      id: "album 01",
      name: "Album 1",
      href: "https://open.spotify.com/album/1",
      images: [
        {
          height: 240,
          width: 240,
          url: "https://mosaic.scdn.co/101",
        }
      ]
    }
  },
  {
    playlistId: databasePlaylists[0].id,
    trackId: "track 01",
    name: "Name 1",
    skippedDate: new Date("2023-10-31T19:52:07Z"),
    artists: [
      {
        id: "artist 01",
        name: "Artist 1",
        href: "https://open.spotify.com/artist/1",
      }
    ],
    album: {
      id: "album 01",
      name: "Album 1",
      href: "https://open.spotify.com/album/1",
      images: [
        {
          height: 240,
          width: 240,
          url: "https://mosaic.scdn.co/101",
        }
      ]
    }
  },
  {
    playlistId: databasePlaylists[0].id,
    trackId: "track 02",
    name: "Name 2",
    skippedDate: new Date("2023-10-31T19:53:10Z"),
    artists: [
      {
        id: "artist 02",
        name: "Artist 2",
        href: "https://open.spotify.com/artist/2",
      }
    ],
    album: {
      id: "album 01",
      name: "Album 2",
      href: "https://open.spotify.com/album/2",
      images: [
        {
          height: 240,
          width: 240,
          url: "https://mosaic.scdn.co/201",
        }
      ]
    }
  },
  {
    playlistId: databasePlaylists[0].id,
    trackId: "track 02",
    name: "Name 2",
    skippedDate: new Date("2023-10-31T19:53:30Z"),
    artists: [
      {
        id: "artist 02",
        name: "Artist 2",
        href: "https://open.spotify.com/artist/2",
      }
    ],
    album: {
      id: "album 01",
      name: "Album 2",
      href: "https://open.spotify.com/album/2",
      images: [
        {
          height: 240,
          width: 240,
          url: "https://mosaic.scdn.co/201",
        }
      ]
    }
  }
];
