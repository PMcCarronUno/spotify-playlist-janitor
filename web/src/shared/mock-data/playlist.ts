import { DetailedPlaylist } from "modules/playlist/playlist-types";
import {
  SkippedTrackHistory,
  SkippedTrackTotal,
  Track,
} from "modules/playlist/tabs/playlist-tabs-types";
import { DEFAULT_ALBUM_IMAGE } from "shared/constants";
import {
  databasePlaylists,
  spotifyPlaylists,
  skippedTracks,
  spotifyTracks as mockSpotifyTracks,
} from "shared/mock-data/api";
import { Image } from "shared/types";

const image: Image = {
  height:
    spotifyPlaylists.find((x) => x.id === databasePlaylists[0].id)?.images[0]
      ?.height || 100,
  width:
    spotifyPlaylists.find((x) => x.id === databasePlaylists[0].id)?.images[0]
      ?.width || 100,
  url:
    spotifyPlaylists.find((x) => x.id === databasePlaylists[0].id)?.images[0]
      ?.url || "href",
};

export const playlist: DetailedPlaylist = {
  id: databasePlaylists[0].id,
  href:
    spotifyPlaylists.find((x) => x.id === databasePlaylists[0].id)?.href || "",
  name:
    spotifyPlaylists.find((x) => x.id === databasePlaylists[0].id)?.name || "",
  image: image,
  ignoreInitialSkips: databasePlaylists[0].ignoreInitialSkips,
  autoCleanupLimit: databasePlaylists[0].autoCleanupLimit,
  skipThreshold: databasePlaylists[0].skipThreshold,
};

export const skippedTrackTotals: SkippedTrackTotal[] = [
  {
    id: skippedTracks[0].trackId,
    name: skippedTracks[0].name,
    duration: skippedTracks[0].duration,
    skippedTotal: 3,
    album: skippedTracks[0].album,
    artists: skippedTracks[0].artists,
    image: skippedTracks[0].album.images.reduce((prev, curr) => {
      return prev.height < curr.height ? prev : curr;
    }),
  },
  {
    id: skippedTracks[3].trackId,
    name: skippedTracks[3].name,
    duration: skippedTracks[3].duration,
    skippedTotal: 2,
    album: skippedTracks[3].album,
    artists: skippedTracks[3].artists,
    image: skippedTracks[3].album.images.reduce((prev, curr) => {
      return prev.height < curr.height ? prev : curr;
    }),
  },
];

export const skippedTrackHistory: SkippedTrackHistory[] = skippedTracks.map(
  (skippedTrack) => ({
    id: skippedTrack.trackId,
    name: skippedTrack.name,
    duration: skippedTrack.duration,
    skippedDate: new Date(skippedTrack.skippedDate),
    album: skippedTrack.album,
    artists: skippedTrack.artists,
    image: skippedTrack.album.images.reduce((prev, curr) => {
      return prev.height < curr.height ? prev : curr;
    }),
  })
);

export const spotifyTracks: Track[] = mockSpotifyTracks.map((spotifytrack) => {
  let image: Image = {
    height: 40,
    width: 40,
    url: DEFAULT_ALBUM_IMAGE,
  };

  if (spotifytrack.album.images.length > 0) {
    image = spotifytrack.album.images.reduce((prev, curr) => {
      return prev.height < curr.height ? prev : curr;
    });
  }

  return {
    id: spotifytrack.id,
    name: spotifytrack.name,
    duration: spotifytrack.duration,
    album: spotifytrack.album,
    artists: spotifytrack.artists,
    image,
  };
});
