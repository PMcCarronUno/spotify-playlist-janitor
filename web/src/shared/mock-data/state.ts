import { UserToken } from "shared/types";

export const userState: UserToken = {
  refresh_token: "access_token",
  token_type: "Bearer",
  access_token: "access_token",
  expires_in: 4503,
  expires_on: new Date().getTime() + 1000 * 100,
};
