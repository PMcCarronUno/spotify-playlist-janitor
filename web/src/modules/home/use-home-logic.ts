import { useCallback, useEffect, useState } from "react";

import { useDataApi } from "api/data-api";
import { useSpotifyApi } from "api/spotify-api";

import { Playlist } from "./home-types";
import { useModal } from "shared/components/modal";

export const useHomeLogic = () => {
  const { getDatabasePlaylists, addDatabasePlaylist } = useDataApi();
  const { getSpotifyPlaylists } = useSpotifyApi();

  const [loading, setLoading] = useState(false);
  const [modalSaving, setModalSaving] = useState(false);
  const [showSpotifyAuthModal, setShowSpotifyAuthModal] = useState(false);
  const [monitoredPlaylists, setMonitoredPlaylists] = useState<
    Playlist[] | undefined
  >(undefined);
  const [unmonitoredPlaylists, setUnmonitoredPlaylists] = useState<
    Playlist[] | undefined
  >(undefined);
  const [modalError, setModalError] = useState<string | undefined>(undefined);

  const { isOpen: modalOpen, onOpen: onModalOpen, onClose } = useModal();

  const onModalClose = useCallback(() => {
    setModalError(undefined);
    onClose();
  }, [setModalError, onClose]);

  const getPlaylistData = useCallback(async () => {
    try {
      setLoading(true);
      const databasePlaylistsResponse = await getDatabasePlaylists();
      const spotifyPlaylistsResponse = await getSpotifyPlaylists();

      setMonitoredPlaylists(
        spotifyPlaylistsResponse.data
          .filter((spotifyPlaylist) =>
            databasePlaylistsResponse.data.some(
              (databasePlaylist) => databasePlaylist.id === spotifyPlaylist.id
            )
          )
          .map((monitoredPlaylist) => ({
            id: monitoredPlaylist.id,
            href: monitoredPlaylist.href,
            name: monitoredPlaylist.name,
            image: monitoredPlaylist.images[0],
          }))
      );

      setUnmonitoredPlaylists(
        spotifyPlaylistsResponse.data
          .filter(
            (spotifyPlaylist) =>
              !databasePlaylistsResponse.data.some(
                (databasePlaylist) => databasePlaylist.id === spotifyPlaylist.id
              )
          )
          .map((unmonitoredPlaylist) => ({
            id: unmonitoredPlaylist.id,
            href: unmonitoredPlaylist.href,
            name: unmonitoredPlaylist.name,
            image: unmonitoredPlaylist.images[0],
          }))
      );
      setLoading(false);
    } catch (e: any) {
      if (
        e?.response?.status === 500 &&
        e?.response?.data?.message ===
          "Application has not been logged into your Spotify account."
      ) {
        setShowSpotifyAuthModal(true);
      }
      setLoading(false);
    }
  }, [
    setLoading,
    getDatabasePlaylists,
    getSpotifyPlaylists,
    setMonitoredPlaylists,
    setUnmonitoredPlaylists,
  ]);

  useEffect(() => {
    getPlaylistData();
  }, [getPlaylistData]);

  const onSubmit = useCallback(
    async (event: React.FormEvent<HTMLFormElement>) => {
      event.preventDefault();
      setModalError(undefined);
      const playlistId = (event.currentTarget[1] as HTMLInputElement).value;
      const skipThreshold = (event.currentTarget[2] as HTMLInputElement).value;
      const ignoreInitialSkips = (event.currentTarget[3] as HTMLInputElement)
        .checked;
      const autoDeleteTracksAfter = (event.currentTarget[4] as HTMLInputElement)
        .value;

      if (!playlistId) {
        setModalError("Please select a playlist");
        return;
      }
      if (!!autoDeleteTracksAfter && parseInt(autoDeleteTracksAfter) === 0) {
        setModalError("Auto-delete must be empty or greater than 0");
        return;
      }

      try {
        setModalSaving(true);
        
        const request = {
          id: playlistId,
          skipThreshold: !!skipThreshold ? parseInt(skipThreshold) : undefined,
          ignoreInitialSkips: ignoreInitialSkips,
          autoCleanupLimit: !!autoDeleteTracksAfter
            ? parseInt(autoDeleteTracksAfter)
            : undefined,
        };

        await addDatabasePlaylist(request);
        setModalSaving(false);
        onModalClose();
        await getPlaylistData();
      } catch (e: any) {
        setModalError(e.response?.data?.message || "Unknown error");
        setModalSaving(false);
      }
    },
    [addDatabasePlaylist, getPlaylistData, onModalClose]
  );

  const onPlaylistChange = useCallback(
    async (newValue: { label: string; value: string }) => {
      if (newValue?.value) {
        setModalError(undefined);
      }
    },
    []
  );

  return {
    monitoredPlaylists,
    unmonitoredPlaylists,
    loading,
    modalOpen,
    onModalOpen,
    onModalClose,
    onSubmit,
    onPlaylistChange,
    modalError,
    modalSaving,
    showSpotifyAuthModal,
  } as const;
};
