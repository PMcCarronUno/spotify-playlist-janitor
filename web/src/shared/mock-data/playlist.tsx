import format from "date-fns/format";
import { DetailedPlaylist } from "modules/playlist/playlist-types";
import {
  SkippedTrackAlbum,
  SkippedTrackArtist,
  SkippedTrackHistory,
  SkippedTrackTotal,
  Track,
} from "modules/playlist/tabs/playlist-tabs-types";
import { PiClock } from "react-icons/pi";
import { VscTrash } from "react-icons/vsc";
import { Skeleton, SkeletonTheme } from "shared/components/skeleton";
import { Column } from "shared/components/table";
import { SecondaryText } from "shared/components/typography";
import {
  DEFAULT_ALBUM_IMAGE,
  SKELETON_HIGHLIGHT,
  SKELETON_PLAYLIST_BASE,
} from "shared/constants";
import {
  databasePlaylists,
  spotifyPlaylists,
  skippedTracks,
  spotifyTracks as mockSpotifyTracks,
} from "shared/mock-data/api";
import { Image } from "shared/types";

const image: Image = {
  height:
    spotifyPlaylists.find((x) => x.id === databasePlaylists[0].id)?.images[0]
      ?.height || 100,
  width:
    spotifyPlaylists.find((x) => x.id === databasePlaylists[0].id)?.images[0]
      ?.width || 100,
  url:
    spotifyPlaylists.find((x) => x.id === databasePlaylists[0].id)?.images[0]
      ?.url || "href",
};

export const playlist: DetailedPlaylist = {
  id: databasePlaylists[0].id,
  href:
    spotifyPlaylists.find((x) => x.id === databasePlaylists[0].id)?.href || "",
  name:
    spotifyPlaylists.find((x) => x.id === databasePlaylists[0].id)?.name || "",
  image: image,
  ignoreInitialSkips: databasePlaylists[0].ignoreInitialSkips,
  autoCleanupLimit: databasePlaylists[0].autoCleanupLimit,
  skipThreshold: databasePlaylists[0].skipThreshold,
};

export const skippedTrackTotals: SkippedTrackTotal[] = [
  {
    id: skippedTracks[0].trackId,
    name: skippedTracks[0].name,
    duration: skippedTracks[0].duration,
    skippedTotal: 3,
    album: skippedTracks[0].album,
    artists: skippedTracks[0].artists,
    image: skippedTracks[0].album.images.reduce((prev, curr) => {
      return prev.height < curr.height ? prev : curr;
    }),
  },
  {
    id: skippedTracks[3].trackId,
    name: skippedTracks[3].name,
    duration: skippedTracks[3].duration,
    skippedTotal: 2,
    album: skippedTracks[3].album,
    artists: skippedTracks[3].artists,
    image: skippedTracks[3].album.images.reduce((prev, curr) => {
      return prev.height < curr.height ? prev : curr;
    }),
  },
];

export const skippedTrackHistory: SkippedTrackHistory[] = skippedTracks.map(
  (skippedTrack) => ({
    id: skippedTrack.trackId,
    name: skippedTrack.name,
    duration: skippedTrack.duration,
    skippedDate: new Date(skippedTrack.skippedDate),
    album: skippedTrack.album,
    artists: skippedTrack.artists,
    image: skippedTrack.album.images.reduce((prev, curr) => {
      return prev.height < curr.height ? prev : curr;
    }),
  })
);

export const spotifyTracks: Track[] = mockSpotifyTracks.map((spotifytrack) => {
  let image: Image = {
    height: 40,
    width: 40,
    url: DEFAULT_ALBUM_IMAGE,
  };

  if (spotifytrack.album.images.length > 0) {
    image = spotifytrack.album.images.reduce((prev, curr) => {
      return prev.height < curr.height ? prev : curr;
    });
  }

  return {
    id: spotifytrack.id,
    name: spotifytrack.name,
    duration: spotifytrack.duration,
    album: spotifytrack.album,
    artists: spotifytrack.artists,
    image,
  };
});

const skeleton = (
  <SkeletonTheme
    baseColor={SKELETON_PLAYLIST_BASE}
    highlightColor={SKELETON_HIGHLIGHT}
    height="100%"
  >
    <Skeleton />
  </SkeletonTheme>
);

export const imageColumn: Column = {
  label: "",
  accessor: "image",
  primary: true,
  render: (image: Image, loading?: boolean) => (
    <div className="centered">
      {loading ? (
        <div style={{ height: 40, width: 40 }}>{skeleton}</div>
      ) : (
        <img alt="" width={40} height={40} src={image.url} />
      )}
    </div>
  ),
};

export const titleColumn: Column = {
  label: "Title",
  accessor: "name",
  sortable: true,
  primary: true,
  render: (value: string, loading?: boolean) => (loading ? skeleton : value),
};

export const artistColumn: Column = {
  label: "Artist",
  accessor: "artists",
  sortable: true,
  render: (artists: SkippedTrackArtist[], loading?: boolean) => (
    <>
      {loading
        ? skeleton
        : artists.map((artist, index) => (
            <>
              {`${index === 0 ? "" : ", "}`}
              {!!artist.href ? (
                <a href={artist.href} target="_blank" rel="noreferrer">
                  {artist.name}
                </a>
              ) : (
                <SecondaryText>{artist.name}</SecondaryText>
              )}
            </>
          ))}
    </>
  ),
};

export const albumColumn: Column = {
  label: "Album",
  accessor: "album",
  sortable: true,
  render: (album: SkippedTrackAlbum, loading?: boolean) =>
    loading ? (
      skeleton
    ) : !!album.href ? (
      <a href={album.href} target="_blank" rel="noreferrer">
        {album.name}
      </a>
    ) : (
      <SecondaryText>{album.name}</SecondaryText>
    ),
};

export const durationColumn: Column = {
  label: <PiClock />,
  accessor: "duration",
  sortable: true,
  render: (millis: number, loading?: boolean) => {
    if (loading) {
      return skeleton;
    }
    var minutes = Math.floor(millis / 60000);
    var seconds = parseInt(((millis % 60000) / 1000).toFixed(0));
    return seconds === 60
      ? minutes + 1 + ":00"
      : minutes + ":" + (seconds < 10 ? "0" : "") + seconds;
  },
};

export const totalSkippedColumn: Column = {
  label: "Date Skipped",
  accessor: "skippedDate",
  sortable: true,
  sortbyOrder: "desc",
  render: (date: Date, loading?: boolean) =>
    loading
      ? skeleton
      : format(date, "MMM dd yyyy") + " at " + format(date, "HH:mm:ss"),
};

export const dateSkippedColumn: Column = {
  label: "Total Skipped",
  accessor: "skippedTotal",
  sortable: true,
  sortbyOrder: "desc",
  render: (value: string, loading?: boolean) => (loading ? skeleton : value),
};

export const getDeleteColumn = (onDeleteClick: (id: string) => void) => ({
  label: "",
  accessor: "id",
  render: (id: string, loading?: boolean) =>
    loading ? (
      <></>
    ) : (
      <div className="centered">
        <VscTrash
          style={{ cursor: "pointer" }}
          onClick={() => onDeleteClick(id)}
        />
      </div>
    ),
});
