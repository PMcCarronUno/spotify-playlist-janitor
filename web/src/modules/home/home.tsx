import styled from "styled-components";
import { AiOutlinePlusCircle } from "react-icons/ai";

import { BLACK, GREEN, RED } from "shared/constants";
import { SubTitle, Text } from "shared/components/typography";
import { Modal } from "shared/components/modal";

import { useHomeLogic } from "./use-home-logic";
import { AddPlaylistModalView } from "./modal/add-playlist-modal";

export const Home = () => {
  const {
    monitoredPlaylists,
    unmonitoredPlaylists,
    loading,
    modalOpen,
    onModalOpen,
    onModalClose,
    onSubmit,
    onPlaylistChange,
    modalError,
    modalSaving,
    showSpotifyAuthModal,
  } = useHomeLogic();

  return (
    <PageContainer>
      {monitoredPlaylists && (
        <>
          {monitoredPlaylists.map((monitoredPlaylist, index) => (
            <div
              key={monitoredPlaylist.id}
              className="item"
              onClick={() =>
                alert(
                  `Naviagte to ${monitoredPlaylist.name} at route ${monitoredPlaylist.id}`
                )
              }
              id={`playlist-item-${index}`}
              data-testid={`playlist-item-${index}`}
            >
              {monitoredPlaylist.image && (
                <img
                  alt={monitoredPlaylist.name}
                  width={130}
                  height={130}
                  src={monitoredPlaylist.image.url}
                />
              )}

              <SubTitle>{monitoredPlaylist.name}</SubTitle>
            </div>
          ))}
          <div
            className="item new"
            onClick={onModalOpen}
            id={`add-playlist-item`}
            data-testid={`add-playlist-item`}
          >
            <AiOutlinePlusCircle />
          </div>
        </>
      )}
      <Modal
        {...{
          isOpen: modalOpen,
          onClose: onModalClose,
          label: "Select Playlist Modal",
        }}
      >
        <AddPlaylistModalView
          {...{
            onSubmit,
            unmonitoredPlaylists,
            onPlaylistChange,
            modalSaving,
            modalError,
            onModalClose,
          }}
        />
      </Modal>
      <Modal
        {...{
          isOpen: showSpotifyAuthModal,
          onClose: () => {},
          label: "Spotify Auth Modal",
        }}
      >
        <ModalContainer>
          <SubTitle style={{ marginBottom: 8 }}>
            Spotify Authentication Error
          </SubTitle>
          <Text>
            The application has not been authenticated with your Spotify
            account. Please follow{" "}
            <a
              href={process.env.REACT_APP_API_URL}
              target="_blank"
              rel="noreferrer"
            >
              this
            </a>{" "}
            link and sign into Spotify, and <a href=".">reload</a> this page.
          </Text>
        </ModalContainer>
      </Modal>
    </PageContainer>
  );
};

const PageContainer = styled.div`
  display: flex;
  flex-wrap: wrap;
  align-items: center;
  align-content: flex-start;

  color: white;
  width: 100%;
  max-width: 1825px;
  paddingtop: 18px;

  .item {
    flex: 1 1 30%; /*grow | shrink | basis */
    display: flex;
    justify-content: space-evenly;
    align-items: center;
    margin: 8px;
    height: 160px;
    max-width: 33%;
    background-color: ${BLACK};

    border-radius: 15px;

    cursor: pointer;

    &:hover {
      transform: scale(1.01);
      box-shadow: 0 0 32px rgba(117, 117, 117, 0.2);
    }
  }

  .new {
    background-color: ${GREEN};

    svg {
      height: 6em;
      width: 6em;
    }
  }
`;

const ModalContainer = styled.div`
  width: 360px;

  .rows {
    margin-bottom: 8px;
  }

  .row {
    display: flex;
    flex-direction: row;
    align-items: center;
    margin-bottom: 8px;
  }

  .number-input {
    width: 55px !important;
  }

  .toggle-input {
    margin-right: 16px;
  }

  .auto-delete-input {
    margin-right: -10px;
  }

  .auto-delete-post-text {
    margin-right: 10px;
  }

  .error {
    color: ${RED};
  }
`;

Home.displayName = "Home";
