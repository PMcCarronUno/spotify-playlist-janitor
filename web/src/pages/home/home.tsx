import styled from "styled-components";
import { HEADER_HEIGHT, HEADER_PADDING } from "shared/constants";
import { Home as HomeInitialise } from "modules/home";

export const Home = () => {
  return (
    <Content>
      <HomeInitialise />
    </Content>
  );
};

const Content = styled.div`
  display: flex;
  justify-content: center;
  height: calc(
    100vh - ${HEADER_HEIGHT} - ${HEADER_PADDING} - ${HEADER_PADDING}
  );
`;

Home.displayName = "Home";
