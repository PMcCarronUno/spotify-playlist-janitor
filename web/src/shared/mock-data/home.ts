import { Playlist } from "modules/home/home-types";
import { databasePlaylists, spotifyPlaylists } from "shared/mock-data/api";

export const monitoredPlaylists: Playlist[] = spotifyPlaylists
  .filter((spotifyPlaylist) =>
    databasePlaylists.some(
      (databasePlaylist) => databasePlaylist.id === spotifyPlaylist.id
    )
  )
  .map((monitoredPlaylist) => ({
    id: monitoredPlaylist.id,
    href: monitoredPlaylist.href,
    name: monitoredPlaylist.name,
    image: monitoredPlaylist.images[0],
  }));

export const unmonitoredPlaylists: Playlist[] = spotifyPlaylists
  .filter(
    (spotifyPlaylist) =>
      !databasePlaylists.some(
        (databasePlaylist) => databasePlaylist.id === spotifyPlaylist.id
      )
  )
  .map((monitoredPlaylist) => ({
    id: monitoredPlaylist.id,
    href: monitoredPlaylist.href,
    name: monitoredPlaylist.name,
    image: monitoredPlaylist.images[0],
  }));
