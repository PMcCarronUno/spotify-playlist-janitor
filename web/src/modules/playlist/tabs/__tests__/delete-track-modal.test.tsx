import { fireEvent, getByTestId, render } from "@testing-library/react";

import { DeleteTrackModalView } from "../delete-track-modal";
import { unmonitoredPlaylists } from "shared/mock-data/home";

describe("<DeleteTrackModalView />", () => {
  let container: HTMLElement;
  let props: any = {
    onSubmit: jest.fn(),
    deleteTracks: [{ id: "track_id", name: "Track Name" }],
    modalSaving: false,
    modalError: undefined,
    onModalClose: jest.fn(),
  };

  beforeEach(async () => {
    ({ container } = render(<DeleteTrackModalView {...props} />));
  });

  it("should render DeleteTrackModalView component", () => {
    expect(container.firstChild).toMatchSnapshot();
  });

  it("should render DeleteTrackModalView component with an error", () => {
    props = {
      ...props,
      modalError: "MODAL_ERROR",
    };

    ({ container } = render(<DeleteTrackModalView {...props} />));

    expect(container.firstChild).toMatchSnapshot();
  });

  it("should call onSubmit function when Confirm button clicked", () => {
    const confirm = getByTestId(container, "confirm-modal-button");
    fireEvent.click(confirm);

    expect(props.onSubmit).toHaveBeenCalled();
  });

  it("should call onModalClose function when Cancel button clicked", () => {
    const cancel = getByTestId(container, "cancel-modal-button");
    fireEvent.click(cancel);

    expect(props.onModalClose).toHaveBeenCalled();
  });
});
