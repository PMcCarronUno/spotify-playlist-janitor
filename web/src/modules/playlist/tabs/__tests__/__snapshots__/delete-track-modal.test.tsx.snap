// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<DeleteTrackModalView /> should render DeleteTrackModalView component 1`] = `
<form
  autocomplete="off"
>
  <div
    class="sc-fsYfxw jIlICe"
  >
    <div
      class="sc-guDMob jLqGJD"
      style="margin-bottom: 8px;"
    >
      Delete Track
    </div>
    <div
      class="sc-dmyDmy dcXriK"
      style="margin-bottom: 8px;"
    >
      Are you sure you want to delete the following track from this playlist?
    </div>
    <ul>
      <li>
        Track Name
      </li>
    </ul>
    <div
      class="rows"
    />
    <div
      style="display: flex; flex-direction: row; justify-content: flex-end;"
    >
      <button
        class="sc-jTQDJr sc-iBdoyZ gmttQv bbczBv secondary"
        data-testid="cancel-modal-button"
        id="cancel-modal-button"
        style="margin: 0px 4px;"
        type="button"
      >
        Cancel
      </button>
      <button
        class="sc-jTQDJr sc-gLLuHO gmttQv lckOnr primary"
        data-testid="confirm-modal-button"
        id="confirm-modal-button"
        style="margin: 0px 4px;"
        type="submit"
      >
        Confirm
      </button>
    </div>
  </div>
</form>
`;

exports[`<DeleteTrackModalView /> should render DeleteTrackModalView component with an error 1`] = `
<form
  autocomplete="off"
>
  <div
    class="sc-fsYfxw jIlICe"
  >
    <div
      class="sc-guDMob jLqGJD"
      style="margin-bottom: 8px;"
    >
      Delete Track
    </div>
    <div
      class="sc-dmyDmy dcXriK"
      style="margin-bottom: 8px;"
    >
      Are you sure you want to delete the following track from this playlist?
    </div>
    <ul>
      <li>
        Track Name
      </li>
    </ul>
    <div
      class="rows"
    >
      <div
        class="row error"
        data-testid="modal-error"
        id="modal-error"
      >
        <div
          class="sc-hLQTFJ gjIzVE"
        >
          MODAL_ERROR
        </div>
      </div>
    </div>
    <div
      style="display: flex; flex-direction: row; justify-content: flex-end;"
    >
      <button
        class="sc-jTQDJr sc-iBdoyZ gmttQv bbczBv secondary"
        data-testid="cancel-modal-button"
        id="cancel-modal-button"
        style="margin: 0px 4px;"
        type="button"
      >
        Cancel
      </button>
      <button
        class="sc-jTQDJr sc-gLLuHO gmttQv lckOnr primary"
        data-testid="confirm-modal-button"
        id="confirm-modal-button"
        style="margin: 0px 4px;"
        type="submit"
      >
        Confirm
      </button>
    </div>
  </div>
</form>
`;
